### DISCUSSION TITLE
`RequestValidationError` is raised when testing, instead of returing `422` response

### DISCUSSION THREAD

**Discussion by jkazimierczak** [fastapi/fastapi#6316] [Questions] [answered]
### First Check

- [X] I added a very descriptive title to this issue.
- [X] I used the GitHub search to find a similar issue and didn't find it.
- [X] I searched the FastAPI documentation, with the integrated search.
- [X] I already searched in Google "How to X in FastAPI" and didn't find any information.
- [X] I already read and followed all the tutorial in the docs and didn't find an answer.
- [X] I already checked if it is not related to FastAPI but to [Pydantic](https://github.com/samuelcolvin/pydantic).
- [X] I already checked if it is not related to FastAPI but to [Swagger UI](https://github.com/swagger-api/swagger-ui).
- [X] I already checked if it is not related to FastAPI but to [ReDoc](https://github.com/Redocly/redoc).

### Commit to Help

- [X] I commit to help with one of those options 👆

### Example Code

```python
import uuid

from fastapi import FastAPI, APIRouter, Header, HTTPException
from fastapi.testclient import TestClient
from pydantic import BaseModel


# Models and CRUD
class Item(BaseModel):
    id: int
    content: str


def crud_get_items():
    # This function lies in a separate module
    return [
        Item(id=1, content="First item"),
        Item(id=2, content="Second item"),
    ]


# App
app = APIRouter()


@app.get("/")
async def get_items(user_id: str = Header()):
    data = crud_get_items()
    if not len(data):  # This is redundant in the example, but exist in "normal" codebase
        raise HTTPException(status_code=404, detail="No items found")

    return data


# Tests
client = TestClient(app)


def test_get_items():
    headers = {"user-id": str(uuid.uuid4())}
    res = client.get("/", headers=headers)
    assert res.status_code == 200


def test_get_items_without_userid():
    res = client.get("/")
    assert res.status_code == 422
```

### Description

When testing my app I noticed, that sending request to router without header parameter set, I get `RequestValidationError` instead of `422` status code. **This doesn't happen, when using `FastAPI` instead of `APIRouter`**.

One could say, that an easy fix to this is to wrap the second test code in `pytest.raises()`, but this will only execute the `client.get` while skipping all asserts the test has.
```py
from fastapi.exceptions import RequestValidationError

def test_get_items_without_userid():
    with pytest.raises(RequestValidationError):
        res = client.get("/")          # This line will execute
        assert res.status_code == 422  # This line won't (and any below) 
```

### Operating System

Windows

### Operating System Details

Windows 10, 21H2

### FastAPI Version

0.85.0

### Python Version

Python 3.10.4

**Comment by iudeen** [comment]
Why are you not initializing the app?

Like `app = FastAPI()`?

**Comment by jkazimierczak** [comment]
In production code I'm using it. The API itself is mounted under `/api`, as I'm serving static files from "/".

So to avoid the need to write `/api/items/<anything that comes after that>` I thought that maybe the TestClient has some way to set the base (`/api/items`) for each request. I tried passing `base_url` and `root_path` like so:
```py
# One at the time, of course
client = TestClient(app, base_url="http://testserver/api")
client = TestClient(app, root_path="/api")
```
but this didn't change anything. So to avoid repetitive writing, testing the router itself seemed like a good solution.

If routers aren't meant to be tested directly, how one can shorten request base path (in my case `/api/items`)?

**Comment by jkazimierczak** [comment]
Now when I wrote this, I thought of solution - initialize app per test file, like so:
```py
# all imports here

app = FastAPI()
app.include_router(router)  # router I want to test
client = TestClient(app)

# tests below
```

Then it works correctly. But at this point I want to know if `APIRouter` is untestable on it's own by design?

**Comment by JarroVGIT** [comment, marked as answer]
The `TestClient` is created to test `Starlette` applications, and by extension you can use it to test `FastAPI` applications. It is not designed to test Starlette's `Router` or FastAPI's `APIRouter` directly. As you have experienced, it can yield unexpected results. Most notably: The difference between sending a request to a `FastAPI` application and sending a request to the `APIRouter`  directly (using it as an ASGI application on its own), is that no middleware is executed when doing the latter.

And by default (and design), there are always at least 2 middleware present in your FastAPI application (which will not be present in your `APIRouter` instance: one that catches and handles exceptions like `HTTPException` and one (the very last one) that handles any uncaught exception and returns a `HTTP 500 Internal Server Error` to ensure your server keeps running. 

So, when using `APIRouter` directly in your `TestClient`, any raised `HTTPException`s will not result in a proper response, rather than actually raised (and unhandled) exception.

So, to answer your question:
> But at this point I want to know if APIRouter is untestable on it's own by design?

The behaviour that you are experiencing is by design, and when testing an `APIRouter` one should keep in mind that `FastAPI + APIRouter` has different behaviour than just the `APIRouter` as an ASGI application on its own.

**Comment by jkazimierczak** [comment]
Thank you for very through explanation of the differences between `FastAPI` and `APIRouter`. This is very helpful and insightful. It makes sense now why I was was getting exceptions and not the responses with correct status codes.

**Comment by tiangolo** [comment]
Thanks for the help here @JarroVGIT ! 👏  You rock!  🙇 

Thanks for reporting back and closing the issue @jkazimierczak  👍
